<a href="https://colab.research.google.com/github/Gr1lledChee5e/OthersDataStuff/blob/main/TFIDF%3A%20clothes_recommendation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Content-based recommender systems
### We will use tf-idf and cosine similarity to recommend the N most similar items
### tf-idf features are the method to convert the textual information into the vector space, they are a measure of how important a word in a text is
### The data set contains 500 entries of different items like shoes, shirts etc., along with an item-id and a textual description of the item.

In [ ]:
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.metrics.pairwise import linear_kernel

In [ ]:
ds = pd.read_csv("https://raw.githubusercontent.com/wangx346/MAS651/main/clothes_review.csv")

In [ ]:
ds.head(10)

,id,description
0,1,Active classic boxers - There's a reason why o...
1,2,Active sport boxer briefs - Skinning up Glory ...
2,3,Active sport briefs - These superbreathable no...
3,4,"Alpine guide pants - Skin in, climb ice, switc..."
4,5,"Alpine wind jkt - On high ridges, steep ice an..."
5,6,Ascensionist jkt - Our most technical soft she...
6,7,"Atom - A multitasker's cloud nine, the Atom pl..."
7,8,Print banded betina btm - Our fullest coverage...
8,9,Baby micro d-luxe cardigan - Micro D-Luxe is a...
9,10,Baby sun bucket hat - This hat goes on when th...


##### The TF*IDF algorithm is used to weigh a keyword in any document and assign the importance to that keyword based on the number of times it appears in the document. Put simply, the higher the TF*IDF score (weight), the more important the term.

In Python, scikit-learn provides you a pre-built TF-IDF vectorizer that calculates the
TF-IDF score for each document’s description, word-by-word.

min_df is used for removing terms that appear too infrequently. For example:
min_df = 0.01 means "ignore terms that appear in less than 1% of the documents".
min_df = 5 means "ignore terms that appear in less than 5 documents".
The default min_df is 1, which means "ignore terms that appear in less than 1 document". Thus, the default setting does not ignore any terms.


max_df is used for removing terms that appear too frequently, also known as "corpus-specific stop words". For example:

max_df = 0.50 means "ignore terms that appear in more than 50% of the documents".
max_df = 25 means "ignore terms that appear in more than 25 documents".
The default max_df is 1.0, which means "ignore terms that appear in more than 100% of the documents". Thus, the default setting does not ignore any terms.

ngram_range: An n-gram is just a string of n words in a row. E.g. the sentence ‘I am Groot’ contains the 2-grams ‘I am’ and ‘am Groot’. The sentence is itself a 3-gram. Set the parameter ngram_range=(a,b) where a is the minimum and b is the maximum size of ngrams you want to include in your features. The default ngram_range is (1,1).

In [ ]:
## instantiate the vectorizer object
tf = TfidfVectorizer(analyzer='word', ngram_range=(1, 3), min_df=0, stop_words='english')
## convert the documents into a matrix
tfidf_matrix = tf.fit_transform(ds['description'])
type(tfidf_matrix)

scipy.sparse.csr.csr_matrix

In [ ]:
tfidf_matrix.shape

(500, 52262)

##### Now we have a ‘tfidf’ feature matrix for all the 500 items. Every item has 52262 number of features (words ).
Here, the tfidf_matrix is the matrix containing each word and its TF-IDF score with regard to each document, or item in this case. Also, stop words are simply words that add no significant value to our system, like ‘an’, ‘is’, ‘the’, and hence are ignored by the system.
Now, we have a representation of every item in terms of its description. Next, we need to calculate the relevance or similarity of one document to another.

## Calculate cosine similarity
Cosine_Similarity is basically a measure of the similarity between 2 vectors. This measure is the cosine of the angle between them. Here, we have 52262 features (tfidf values) for each item.
Here we’ve calculated the cosine similarity of each item with every other item in the dataset, and then arranged them according to their similarity with item i, and stored the values in results.

In [ ]:
# calculating cosine similarity between items

cosine_similarities = cosine_similarity(tfidf_matrix, tfidf_matrix)
cosine_similarities.shape

(500, 500)

In [ ]:
cosine_similarities

array([[1.        , 0.10110642, 0.06487353, ..., 0.06097409, 0.06546914,
        0.06955608],
       [0.10110642, 1.        , 0.4181664 , ..., 0.03550042, 0.06936414,
        0.06480538],
       [0.06487353, 0.4181664 , 1.        , ..., 0.03402428, 0.0455137 ,
        0.05038512],
       ...,
       [0.06097409, 0.03550042, 0.03402428, ..., 1.        , 0.04187121,
        0.04958298],
       [0.06546914, 0.06936414, 0.0455137 , ..., 0.04187121, 1.        ,
        0.36281626],
       [0.06955608, 0.06480538, 0.05038512, ..., 0.04958298, 0.36281626,
        1.        ]])

In [ ]:
type(cosine_similarities)

numpy.ndarray

In [ ]:
cosine_similarities[0].argsort()

array([331, 489, 318, 424, 433,  68,  48, 438, 370, 133,  53,  49,  47,
       351, 395,   9,  82,  66, 237, 136,  41, 348, 421,  85, 430, 434,
       454, 196, 467,  84, 113, 420, 211, 273, 393,  16, 326, 134, 149,
       381,  30,  32, 455, 485, 112, 425, 240,  50,  93, 354, 366, 153,
       152, 252, 243, 337,  12,  98, 129, 132, 423,  78, 392, 419, 344,
        83, 157, 429, 128,  79, 428, 141, 456, 280, 130,  52,  95, 341,
       135, 219, 404, 220, 338,   7, 221, 241, 343, 448, 353, 463, 350,
       106, 110, 315, 289, 367, 274, 140, 397, 299, 306, 270, 436, 260,
       251, 242, 147,  38,  73, 388, 417,  80, 253, 292, 464, 191, 278,
       281, 382, 154, 122, 190,  92,  37,   6, 142, 197, 312, 409, 363,
       214, 383, 209, 246, 148, 303, 144, 313, 225, 471, 290, 218, 250,
       262, 255, 435, 377, 181,  77, 161, 120, 298, 378, 224, 473, 231,
       257, 477, 198, 375, 478, 217, 100, 139, 297, 102, 178, 400, 189,
       232,   5, 131,  43, 239, 254, 472,  34, 261, 182, 205, 29

In [ ]:
## find the top 99 highest cosine similarities
## argsort(): Returns the indices of elements sorted in ascending order (from lowest to highest similarity).
## [:-100:-1] Slicing operation to select the top 99 highest similarities (excluding self-similarity if needed).
## :-100 → Selects the last 99 elements (highest values).


array([  0,  18, 493,  17, 171, 441, 170,  20, 494,  24, 495, 486,  19,
       340, 175, 487, 364, 339,  59, 439, 496, 172, 440, 412, 442, 173,
       358,  21,  60, 311,  22,   1, 359, 174, 328,  23, 405, 265, 276,
       213, 235, 445, 449, 413,  99, 468, 258, 418, 330,  11, 357, 324,
       302, 259, 146, 327, 499, 264,  91,  88, 391, 109, 390,  45,  61,
       199, 200,  87, 329, 179, 498,  63, 334,   8, 301, 323, 415,   2,
       203, 163, 325, 201, 406, 465,  86, 275,  89, 482, 208, 481, 480,
        36,  33, 414,  56, 104, 137, 286, 167], dtype=int64)

#### The function below first determines the index of the input item, retrieves the similarities of itmes with selected item, sorts them and returns the titles of items with the highest similarity to the selected item.

iterrows() is used to iterate over a pandas Data frame rows in the form of (index, series) pair.

argsort(), returns the indices which would sort an array. x = np.array([3, 1, 2]) np.argsort(x) array([1, 2, 0])

In [ ]:
##  when you do a[::-1], it starts from the end towards the first taking each element. So it reverses a.

results = {}

for idx, row in ds.iterrows():
    ## find the top 99 highest cosine similarities
    similar_indices = cosine_similarities[idx].argsort()[:-100:-1]
    ## the the items corresponding to the highest similarities
    similar_items = [(cosine_similarities[idx][i], ds['id'][i]) for i in similar_indices]

    results[row['id']] = similar_items[1:]

print('done!')
type(results)

done!


dict

In [ ]:
results[500][:7]

[(0.3628162618600105, 499),
 (0.31804645992911146, 462),
 (0.31778345313004575, 463),
 (0.31556134422934073, 32),
 (0.2566286733846444, 34),
 (0.23425464417479275, 483),
 (0.2166078992853787, 303)]

## Make recommendation
Here, we just input an item_id and the number of recommendations that we want. Our function collects the results[] corresponding to that item_id, and we get our recommendations on screen.

In [ ]:
def item(id):
    return ds.loc[ds['id'] == id]['description'].tolist()[0].split(' - ')[0]

# Just reads the results out of the dictionary.
def recommend(item_id, num):
    print("Recommending " + str(num) + " products similar to " + item(item_id) + "...")
    print("-------")
    recs = results[item_id][:num]
    for rec in recs:
        print("Recommended: " + item(rec[1]) + " (score:" + str(rec[0]) + ")")

Here’s a glimpse of what happens when you call the above function.

In [ ]:
recommend(item_id=30, num=10)

Recommending 10 products similar to Cotton board shorts...
-------
Recommended: Wavefarer board shorts-21 in. (score:0.14044018978564404)
Recommended: Minimalist board shorts-19 in. (score:0.13793658801766598)
Recommended: Twenty-three's board shorts (score:0.1336824469134404)
Recommended: Paddler board shorts (score:0.13331764205970856)
Recommended: Duck shorts (score:0.1158622466782414)
Recommended: Custodian pants (score:0.10690842537522768)
Recommended: All-wear shorts (score:0.10526227846780732)
Recommended: Light and variable surf trunks (score:0.10502889619007406)
Recommended: Custodian pants (score:0.10459695983928953)
Recommended: Custodian pants (score:0.10447441847496913)
